subprojects/analysis-markdown-jb/src/main/kotlin/org/jetbrains/dokka/analysis/markdown/jb/MarkdownParser.kt

```diff
@@ -77,6 +77,28 @@ public open class MarkdownParser(
             ).flatMap { it.children }
         )
 
+    /**
+     * Handler for [MarkdownTokenTypes.ATX_CONTENT], which is the content of the header
+     * elements like [MarkdownElementTypes.ATX_1], [MarkdownElementTypes.ATX_2] and so on.
+     *
+     * For example, a header line like `# Header text` is expected to be parsed into:
+     * - One [MarkdownTokenTypes.ATX_HEADER] with startOffset = 0, endOffset = 1 (only the `#` symbol)
+     * - Composite [MarkdownTokenTypes.ATX_CONTENT] with four children: WHITE_SPACE, TEXT, WHITE_SPACE, TEXT.
+     */
+    private fun headerContentHandler(node: ASTNode): List<DocTag> {
+        // ATX_CONTENT contains everything after the `#` symbol, so if there's a space
+        // in-between the `#` symbol and the text (like `# header`), it will be present here too.
+        // However, we don't need the first space between the `#` symbol and the text,
+        // so we just skip it (otherwise the header text will be parsed as `<whitespace>header` instead of `header`).
+        val textStartsWithWhitespace = node.children.firstOrNull()?.type == MarkdownTokenTypes.WHITE_SPACE
+        val children = if (textStartsWithWhitespace) node.children.subList(1, node.children.size) else node.children
```

Comment:
Some tests started failing because they expected `Header text`, but got ` Header text` (notice the whitespace in the beginning).

Quick debugging showed that **before** the update `# header` used to return the following nodes:
* ATX_HEADER
* ATX_CONTENT
    * TEXT

After the update, `# header` started returning
* ATX_HEADER
* ATX_CONTENT
    * WHITE_SPACE
    * TEXT

<details>

<summary>Debugger screenshots</summary>

## Before
 
![image](https://github.com/Kotlin/dokka/assets/22685910/43aae4ad-3173-4962-a2a7-2c83a147d149)

## After

![image](https://github.com/Kotlin/dokka/assets/22685910/703b6d09-f3fc-410f-8f68-bff32ad76ac1)

</details>

Maybe there's a nicer or a better way to fix it - I honestly didn't spend much time on it.

Reply:
Just interesting, how it worked before/after if there are multiple spaces in the beginning? f.e. `#   header` - may be we need to filter all leading white space tokens here?

Reply:
Excellent thought, I'll check :+1:

Reply:
Looks like it's how I expected - there's only one `WHITE_SPACE` element, but it's wider, so there's no need to do anything additionally.

I've added a test and a comment about it though. Also created #3233 after seeing a single test file with 1.6k lines of code :sweat_smile: 

![2023-10-18_12-18-27](https://github.com/Kotlin/dokka/assets/22685910/ef50122b-4ab1-444e-81a2-78889f0dabe6)